Jackson 2.15.0 - Error deserializing an @JsonUnwrapped object with a field of type float

This works as of Jackson 2.14.2, and fails with Jackson 2.15.0, both using bson4jackson 2.13.1

Example class being deserialized: (groovy syntax)
```
static class OuterClass {
    @JsonUnwrapped
    public InnerClass inner = new InnerClass()
    OuterClass() {}
}

static class InnerClass {
    public float floatValue = 40.0f
    public InnerClass() {}
}
```

Failing test (but passes with older jackson):
```
OuterClass outer = new OuterClass()
outer.inner.floatValue = 50.0f
ByteArrayOutputStream baos = new ByteArrayOutputStream()
ObjectMapper BSON_MAPPER = new ObjectMapper(new BsonFactory())
BSON_MAPPER.writeValue(baos, outer)
ByteArrayInputStream bais = new ByteArrayInputStream(baos.toByteArray())
OuterClass clone = BSON_MAPPER.readValue(bais, OuterClass.class)
assertThat(clone.inner.floatValue).isEqualTo(outer.inner.floatValue)
```

I am uncertain if the actual bug exists in jackson or bson4jackson, but the following test using only jackson passes for both old and new versions: (This will use a ReaderBasedJsonParser instead of a BsonParser)
```
ObjectMapper MAPPER = new ObjectMapper()
OuterClass outer = MAPPER.readValue("{\"floatValue\":50.0}", OuterClass.class)
assertThat(outer.inner.floatValue).isEqualTo(50.0f)
```

One relevant change in jackson is in `TokenBuffer._copyBufferValue(JsonParser p, JsonToken t)`
[2.14 version](https://github.com/FasterXML/jackson-databind/blob/d47d1b642931246770ec67c4a98ad6aacc0f7d7d/src/main/java/com/fasterxml/jackson/databind/util/TokenBuffer.java#L1243) vs [2.15 version](https://github.com/FasterXML/jackson-databind/blob/df541d3c74133496454dd9bfccc404b54a0faea6/src/main/java/com/fasterxml/jackson/databind/util/TokenBuffer.java#L1274)
2.14 version calls JsonParser.getNumberValueExact() which was overridden in BsonParser, which gets the value out of BsonParser.Context.value 
2.15 version calls JsonParser.getNumberValueDeferred() which is not implemented by BsonParser, so it calls ParserBase.
getNumberValueDeferred(), which gets the value from _numberString, _numberFloat, _textBuffer as needed.

```
com.fasterxml.jackson.databind.JsonMappingException: empty String (through reference chain: com.apptegic.excavate.util.JacksonUnwrappedTest$InnerClass["floatValue"])

	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:402)
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:361)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.wrapAndThrow(BeanDeserializerBase.java:1830)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:394)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:185)
	at com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:138)
	at com.fasterxml.jackson.databind.deser.impl.UnwrappedPropertyHandler.processUnwrapped(UnwrappedPropertyHandler.java:62)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeWithUnwrapped(BeanDeserializer.java:765)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:347)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:185)
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4825)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3809)
	
Caused by: java.lang.NumberFormatException: empty String
	at java.base/jdk.internal.math.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:1842)
	at java.base/jdk.internal.math.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.base/java.lang.Double.parseDouble(Double.java:543)
	at com.fasterxml.jackson.core.io.NumberInput.parseDouble(NumberInput.java:388)
	at com.fasterxml.jackson.databind.util.TokenBuffer$Parser.getNumberValue(TokenBuffer.java:1936)
	at com.fasterxml.jackson.databind.util.TokenBuffer$Parser.getNumberValue(TokenBuffer.java:1891)
	at com.fasterxml.jackson.databind.util.TokenBuffer$Parser.getFloatValue(TokenBuffer.java:1847)
	at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$FloatDeserializer.deserialize(NumberDeserializers.java:596)
	at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$FloatDeserializer.deserialize(NumberDeserializers.java:579)
	at com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:138)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:392)
	... 35 more
```

--- a/build.gradle
+++ b/build.gradle
@@ -37,6 +37,8 @@ configurations {
     integrationTest211 { extendsFrom integrationTestBase }
     integrationTest212 { extendsFrom integrationTestBase }
     integrationTest213 { extendsFrom integrationTestBase }
+    integrationTest214 { extendsFrom integrationTestBase }
+    integrationTest215 { extendsFrom integrationTestBase }
 }
 
 String jacksonCore = 'com.fasterxml.jackson.core:jackson-core:'
@@ -46,9 +48,9 @@ String junit = 'junit:junit:4.13.2'
 String mongoDbJavaDriver = 'org.mongodb:mongo-java-driver:3.12.12'
 
 dependencies {
-    implementation jacksonCore + '2.13.0'
-    implementation jacksonAnnotations + '2.13.0'
-    implementation jacksonDatabind + '2.13.0'
+    implementation jacksonCore + '2.15.2'
+    implementation jacksonAnnotations + '2.15.2'
+    implementation jacksonDatabind + '2.15.2'
 
     testImplementation junit
     testImplementation mongoDbJavaDriver
@@ -107,6 +109,14 @@ dependencies {
     integrationTest213 jacksonCore + '2.13.0'
     integrationTest213 jacksonAnnotations + '2.13.0'
     integrationTest213 jacksonDatabind + '2.13.0'
+
+    integrationTest214 jacksonCore + '2.14.3'
+    integrationTest214 jacksonAnnotations + '2.14.3'
+    integrationTest214 jacksonDatabind + '2.14.3'
+
+    integrationTest215 jacksonCore + '2.15.2'
+    integrationTest215 jacksonAnnotations + '2.15.2'
+    integrationTest215 jacksonDatabind + '2.15.2'
 }
 
 jar {
@@ -201,11 +211,20 @@ task doIntegrationTest213(type: Test) {
     classpath = sourceSets.test.output + sourceSets.main.output + configurations.integrationTest213
 }
 
+task doIntegrationTest214(type: Test) {
+    classpath = sourceSets.test.output + sourceSets.main.output + configurations.integrationTest214
+}
+
+task doIntegrationTest215(type: Test) {
+    classpath = sourceSets.test.output + sourceSets.main.output + configurations.integrationTest215
+}
+
 task integrationTest(dependsOn: [ doIntegrationTest21, doIntegrationTest22,
         doIntegrationTest23, doIntegrationTest24, doIntegrationTest25,
         doIntegrationTest26, doIntegrationTest27, doIntegrationTest28,
         doIntegrationTest29, doIntegrationTest210, doIntegrationTest211,
-        doIntegrationTest212, doIntegrationTest213 ])
+        doIntegrationTest212, doIntegrationTest213, doIntegrationTest214,
+        doIntegrationTest215])
 
 check.dependsOn integrationTest
 

--- a/src/main/java/de/undercouch/bson4jackson/BsonParser.java
+++ b/src/main/java/de/undercouch/bson4jackson/BsonParser.java
@@ -711,6 +711,11 @@ public class BsonParser extends ParserBase {
     }
 
     @Override
+    public Object getNumberValueDeferred() throws IOException {
+        return getNumberValue();
+    }
+
+    @Override
     public JsonParser.NumberType getNumberType() {
         if (_currentContext == null) {
             return null;

--- a/src/test/java/de/undercouch/bson4jackson/BsonParserTest.java
+++ b/src/test/java/de/undercouch/bson4jackson/BsonParserTest.java
@@ -1,5 +1,6 @@
 package de.undercouch.bson4jackson;
 
+import com.fasterxml.jackson.annotation.JsonUnwrapped;
 import com.fasterxml.jackson.core.JsonParser;
 import com.fasterxml.jackson.core.JsonToken;
 import com.fasterxml.jackson.databind.DeserializationContext;
@@ -653,5 +654,35 @@ public class BsonParserTest {
         Map<?, ?> data = parseBsonObject(o);
         assertEquals(1, data.size());
         assertEquals("MinKey", data.get("A"));
+    }
+
+    static class OuterClass {
+        @JsonUnwrapped
+        public InnerClass inner = new InnerClass();
+    }
+
+    static class InnerClass {
+        public float floatValue = 40.0f;
+    }
+
+    /**
+     * Test if wrapped floats deserialized. Refers issue #121.
+     * @throws Exception if something goes wrong
+     */
+    @Test
+    public void parseWrappedFloat() throws Exception {
+        OuterClass outer = new OuterClass();
+        outer.inner.floatValue = 50.0f;
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+        ObjectMapper mapper = new ObjectMapper(new BsonFactory());
+
+        mapper.writeValue(baos, outer);
+
+        ByteArrayInputStream bais = new ByteArrayInputStream(baos.toByteArray());
+        OuterClass clone = mapper.readValue(bais, OuterClass.class);
+
+        assertEquals(clone.inner.floatValue, outer.inner.floatValue, 0);
     }
 }

